vcd/resource_vcd_vapp_org_network.go

```diff
@@ -0,0 +1,322 @@
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				Description: "Specifies whether the network resources such as IP/MAC of router will be retained across deployments. Default is false.",
+			},
+			"firewall_enabled": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     true,
+				Description: "firewall service enabled or disabled. Default is true",
+			},
+			"nat_enabled": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     true,
+				Description: "NAT service enabled or disabled. Default is true",
+			},
+		},
+	}
+}
+
+func resourceVappOrgNetworkCreate(d *schema.ResourceData, meta interface{}) error {
+	vcdClient := meta.(*VCDClient)
+	vcdClient.lockParentVapp(d)
+	defer vcdClient.unLockParentVapp(d)
+
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return fmt.Errorf(errorRetrievingOrgAndVdc, err)
+	}
+
+	vapp, err := vdc.GetVAppByName(d.Get("vapp_name").(string), false)
+	if err != nil {
+		return fmt.Errorf("error finding vApp. %#v", err)
+	}
+
+	vappNetworkSettings := &govcd.VappNetworkSettings{
+		NatEnabled:         takeBoolPointer(d.Get("nat_enabled").(bool)),
+		FirewallEnabled:    takeBoolPointer(d.Get("firewall_enabled").(bool)),
+		RetainIpMacEnabled: takeBoolPointer(d.Get("retain_ip_mac_enabled").(bool)),
+	}
+
+	orgNetwork, err := vdc.GetOrgVdcNetworkByNameOrId(d.Get("org_network_name").(string), true)
+	if err != nil {
+		return err
+	}
+
+	vAppNetworkConfig, err := vapp.AddOrgNetwork(vappNetworkSettings, orgNetwork.OrgVDCNetwork, d.Get("is_fenced").(bool))
+	if err != nil {
+		return fmt.Errorf("error creating vApp org network. %#v", err)
+	}
+
+	vAppNetwork := types.VAppNetworkConfiguration{}
+	for _, networkConfig := range vAppNetworkConfig.NetworkConfig {
+		if networkConfig.NetworkName == orgNetwork.OrgVDCNetwork.Name {
+			vAppNetwork = networkConfig
+		}
+	}
+
+	if vAppNetwork == (types.VAppNetworkConfiguration{}) {
+		return fmt.Errorf("didn't find vApp network: %s", d.Get("name").(string))
+	}
+
+	// Parsing UUID from 'https://bos1-vcloud-static-170-210.eng.vmware.com/api/admin/network/6ced8e2f-29dd-4201-9801-a02cb8bed821/action/reset'
+	networkId, err := govcd.GetUuidFromHref(vAppNetwork.Link.HREF, false)
+	if err != nil {
+		return fmt.Errorf("unable to get network ID from HREF: %s", err)
+	}
+	d.SetId(networkId)
+
+	return resourceVappOrgNetworkRead(d, meta)
+}
+
+func resourceVappOrgNetworkRead(d *schema.ResourceData, meta interface{}) error {
+	return genericVappOrgNetworkRead(d, meta, "resource")
+}
+
+func genericVappOrgNetworkRead(d *schema.ResourceData, meta interface{}, origin string) error {
+	vcdClient := meta.(*VCDClient)
+
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return fmt.Errorf(errorRetrievingOrgAndVdc, err)
+	}
+
+	vapp, err := vdc.GetVAppByName(d.Get("vapp_name").(string), false)
+	if err != nil {
+		return fmt.Errorf("error finding Vapp: %#v", err)
```

Comment:
```suggestion
		return fmt.Errorf("error finding Vapp: %s", err)
```
Also, please add the vApp name to the error message

Reply:
changed